### PR TITLE
Segment Tree

### LINKED ISSUE #30: Segment Tree
Write a section on segment trees.

### PR DESCRIPTION
**Pull request by mdclyburn**
Segment tree reference for the hackpack. Please review.

Fixes #30.

### CODE REVIEW

File: structures/segment-tree/segment-tree.tex [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+\begin{itemize}
+    \item Line 1: A single integer, $C$, representing the number of cows Farmer John has data on.
+
+    \item The following lines are placed in $C$ groups detailed as follows:
+    \begin{itemize}
+        \item A single line containing two integers, $t_1$ and $t_2$ representing a time range.
+        Should both of these values be 0, Farmer John would not like a count of cows present at any time to be printed.
+        Otherwise, these values are guaranteed to fall between 6 and 20 (inclusive), representing the times between and including 6AM to 8PM.
+        In this case, the highest number of cows seen in the barn at this time should be printed as given in the sample output.
```
Comment (robertu94) [inline structures/segment-tree/segment-tree.tex]: This should be in the section on the output fomat

File: structures/segment-tree/problems/coming-and-going/coming-and-going.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
+#include <iostream>
+#include <cmath>
+using namespace std;
+
```
Comment (robertu94) [inline structures/segment-tree/problems/coming-and-going/coming-and-going.cpp]: This file needs minification macros

**Comment by mdclyburn** [comment]
I have included a separate source file for the reference code to be included in the condensed hackpack. This was easier to do instead of mangling the problem solution source. Problems aren't in the condensed hackpack, are they?

File: structures/segment-tree/segment-tree.tex (hunk 1 above)
Comment (mdclyburn) [inline structures/segment-tree/segment-tree.tex]: Starts on line 69.

**Comment by robertu94** [comment]
The problem statements are not, but the solutions should be.

File: structures/segment-tree/segment-tree.tex (hunk 1 above)
Comment (robertu94) [inline structures/segment-tree/segment-tree.tex]: I didn't say that the output format section was missing. I was saying that information regarding the output format should be in the output fomat section.

Comment (mdclyburn) [inline structures/segment-tree/segment-tree.tex]: Whoops. Gotcha

**Comment by bcdean** [comment]
It's still a bit vague, when reading this, what exactly a segment tree looks like -- can we maybe include a picture or something a bit more concrete in the description that mentions the key feature of how nodes represent "segments" of a sequential data set?  I've read the example problem a few times and I still can't understand exactly what it's asking for, even looking at the input and output samples -- maybe this can be clarified a bit.  Finally, I think the construction time is O(n) instead of O(n log n).  In the code, maybe make it clear how it would need to be modified if we want to do other types of range queries other than range maxes?  (e.g., range min, range sum)

**Comment by bcdean** [comment]
"a segment tree cannot changed" -> "a segment tree cannot change"

**Comment by mdclyburn** [comment]
@bcdean I have tried to make the problem description and input specification clearer. Failing this attempt, perhaps I can explain it to you in depth so you can help me word it better?

File: structures/segment-tree/problems/coming-and-going/coming-and-going.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+		for(unsigned int j = 0; j < in_barn_times; j++)
+		{
+			unsigned int start, end;
+			cin >> start >> end;
+			for(unsigned int k = start; k <= end; k++)
+			{
+				// #ifdef hackpackpp
+				// account for times 0 to 5 we don't care about with k - 6
+				// #endif
+				headcount[k - 6]++;
```
Comment (robertu94) [inline structures/segment-tree/problems/coming-and-going/coming-and-going.cpp]: Perhaps a `#define START_OF_DAY 6` would help with readability of the code.

File: structures/segment-tree/problems/coming-and-going/coming-and-going.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+				update_segment_tree(segtree, 0, N - 1, k - 6, headcount[k - 6], 0);
+			}
+		}
+
+		// #ifdef hackpackpp
+		// print out cow count if requested as data comes in
+		// #endif
+		if(t1 && t2)
```
Comment (robertu94) [inline structures/segment-tree/problems/coming-and-going/coming-and-going.cpp]: even though `(t1 && t2)` is shorter than `(t1 == 0 && t2 == 0)` it is a bit more readable.

File: structures/segment-tree/segment-tree.tex [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+\subsubsection{Input Format}
+
+\begin{itemize}
+    \item Line 1: A single integer, $C$, representing the number of cows Farmer John has data on.
+
+    \item The following lines are placed in $C$ groups detailed as follows:
+    \begin{itemize}
+        \item A single line containing two integers, $t_1$ and $t_2$ representing a time range for which Farmer John wants a count of cows present.
+        These values are guaranteed to fall between 6 and 20 (inclusive), representing the times between and including 6AM to 8PM, except for the case when both are 0.
```
Comment (robertu94) [inline structures/segment-tree/segment-tree.tex]: perhaps this line could be split into two sub-points.  like this

```tex
\item a line containing two integers $t_1$ and $t_2$ $(t_1,t_2 \in {0, x | 6 \leq x \leq 20 })$ which either will be:
\begin{itemize}
    \item two zeros representing that farmer john will not query this range
    \item the time range that farmer john will query after entering these values
\end{itemize}
```

Comment (robertu94) [inline structures/segment-tree/segment-tree.tex]: Alternatively, the `0 0` case doesn't add to much to the problem algorithmically and complicates the problem.  Perhaps we could just remove it.

**Comment by mdclyburn** [comment]
It can be excluded. Removing it is trivial.

**Comment by mdclyburn** [comment]
@bcdean I've taken some input from @robertu94 into consideration and changed the problem slightly. I have also elaborated on the tree construction a little bit more and also included a couple of graphics showing the layout of the tree in the array as well as an abstract visual in a binary tree format. I will look into a better description on segtree construction tomorrow and notify you again.

**Comment by bcdean** [comment]
Ok, thanks – I’ll wait until I hear from your tomorrow to make another pass.

**Comment by mdclyburn** [comment]
@bcdean I have included some more information on the construction of the segment tree. Perhaps this will help make it even clearer as to how it works like it does. I have previously included information in the code on how to modify the segment tree to support different operations. I added some more information there to help that out. If that is not okay, then I could alternatively write out multiple implementations.

For the condensed hackpack, the included reference segment tree code is one that stores RMQs only. I could add the information there as well or write multiple implementations, but I'm not sure if that's desirable for that medium.

I would like you to look over the section again and see what you think of it. Especially the problem.

**Comment by bcdean** [comment]
I like the picture, although  I might make the bottom row of the tree fully complete, and indicate somehow that this bottom row is the “actual” data set—the tree on top of it serves to hold augmented information about ranges in the data set of varying sizes (all powers of two), so any query can be decomposed into a combination of O(log n) of these and hence answered in O(log n) time.

Construction takes only O(n) time, not O(n log n) time, using the method you’ve outlined.

Finally, for the sample question, why does it need a segment tree?  Can’t you just count the number of cows whose intervals overlap with the target [t1,t2] interval as you go using a simple interval overlap check?

--- Brian

**Comment by mdclyburn** [comment]
@bcdean OK. I've changed some things again. With regards to the sample problem, what do you think the issue is? Is it that I'm not being clear enough? Is it confusing? I realize that it might be a little more difficult to analyze the problem since I explained it to you. What do you think @robertu94 ?

**Comment by mdclyburn** [comment]
Enhancements will be handled by #45 .

**Pull request merged**